File: backend/src/main/java/com/college/portal/model/Student.java
package com.college.portal.model;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * Plain POJO - maps to Firestore 'students' collection.
 * Document ID = Firebase Auth UID.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Student {
    private String uid; // Firebase Auth UID (document ID)
    private String firstName;
    private String lastName;
    private String name; // kept for backward compatibility (display name)
    private String email;
    private String role; // "STUDENT" or "ADMIN"
    private String registerNumber;
    private String department;
    private String batch; // e.g. "2023-2027"
    private Integer year; // current year of study e.g. 1,2,3,4
    private String photoUrl; // Google profile photo URL
    private Long createdAt; // epoch millis
    private Long updatedAt; // epoch millis
}

File: backend-node/routes/student.js
const express = require('express');
const admin = require('firebase-admin');
const authenticate = require('../middleware/auth');

const router = express.Router();
router.use(authenticate);

/**
 * GET /api/student/profile
 */
router.get('/profile', async (req, res) => {
    try {
        const db = admin.firestore();
        const doc = await db.collection('students').doc(req.user.uid).get();

        if (!doc.exists) {
            return res.status(404).json({ error: 'Profile not found' });
        }

        return res.json(doc.data());
    } catch (err) {
        return res.status(500).json({ error: err.message });
    }
});

/**
 * PUT /api/student/profile
 * Partial update — only provided fields are changed.
 */
router.put('/profile', async (req, res) => {
    try {
        const uid = req.user.uid;
        const db = admin.firestore();
        const docRef = db.collection('students').doc(uid);
        const doc = await docRef.get();

        if (!doc.exists) {
            return res.status(404).json({ error: 'Profile not found' });
        }

        const student = doc.data();
        const body = req.body;

        const updates = {};
        if (body.firstName !== undefined) { student.firstName = body.firstName; updates.firstName = body.firstName; }
        if (body.lastName !== undefined) { student.lastName = body.lastName; updates.lastName = body.lastName; }
        if (body.registerNumber !== undefined) updates.registerNumber = body.registerNumber;
        if (body.department !== undefined) updates.department = body.department;
        if (body.batch !== undefined) updates.batch = body.batch;
        if (body.year !== undefined) updates.year = typeof body.year === 'string' ? parseInt(body.year) : body.year;
        if (body.photoUrl !== undefined) updates.photoUrl = body.photoUrl;

        // Update display name from first+last
        let displayName = '';
        const fn = updates.firstName || student.firstName || '';
        const ln = updates.lastName || student.lastName || '';
        displayName = (fn + ' ' + ln).trim();
        updates.name = displayName;
        updates.updatedAt = Date.now();

        await docRef.update(updates);
        const updated = (await docRef.get()).data();
        return res.json(updated);
    } catch (err) {
        return res.status(500).json({ error: 'Failed to update profile: ' + err.message });
    }
});

module.exports = router;
